helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java

```diff
@@ -296,7 +297,7 @@ public void close() {
   // Coordinate baseline recalculation and partial rebalance according to the cluster changes.
   private Map<String, IdealState> computeBestPossibleStates(
       ResourceControllerDataProvider clusterData, Map<String, Resource> resourceMap,
-      final CurrentStateOutput currentStateOutput)
+      final CurrentStateOutput currentStateOutput, RebalanceAlgorithm algorithm)
```

Comment:
Note the algorithm could be updated by cluster config change. Then since it could be asynchronously called, if not passing it, we will need to lock the object during the whole calculation. That lock is very heavy and I want to avoid it.

Reply:
@jiajunwang could you not easily resolve the comment, I noticed GitHub PR review won't notify the reviewers some times.
Now I understand why do you prefer to passing the instance in the method parameter (the value can be changed). However, the purpose of doing so is hidden beneath the code itself (it doesn't self explain IMO if the reader doesn't have a whole picture) and you have to pass the rebalance algorithm instance from method A -> B -> C... Why do you think a lock is heavy? I suppose a re-entrant lock is sufficient and way more readable as it makes other developers aware of the possible race condition immediately.  
A private method like this should be good enough
```
private RebalanceAlgorithm getRebalanceAlgorithm() {
   _reentrantLock.acquire();
   algorithm = getInstance();
   _retentranceLock.release();
}
```

Reply:
A more extreme way is:
1. I notice the rebalance algorithm is one-to-one mapping determined by the preference map.
2. If we make the preference map a concurrent map, and *rebalance algorithm instance* is not a private field.
3. Create the rebalance algorithm instance when it's actually needed (to only make sure partial rebalance and global rebalance share the exact same instance).

Reply:
Sorry that I didn't notice that you cannot reopen the conversation...

Actually, the main reason for this is not preventing a simple race condition, but to prevent the different pipelines (baseline and partial rebalance) or different steps get different algorithm objects.
A small lock surrounding the get method won't help. Because we have more than one get in the whole workflow. Given that saying, the second method won't work as well. Whenever the algorithm object is updated, it will be visible to the next get call. And that is not what we want.

Reply:
I understand your requirement and proposed another idea in the other comment. Hopefully, we could have fewer parameters in the method signature. Let me know what you think

Reply:
Less parameters is, in general, preferred. But that is not the high priority. Correctness is. \
I agree that there could be alternative way to do this. For example, creating other methods with different parameter list or even re-organize the whole class design. But all the ways that I tried have made it even more complicated.

I'm open about any good idea to improve it. But till now, this is the most clean way that I can see.

Reply:
Your concern about creating a new rebalance algorithm instance every time (which can be resource heavy especially it involves property file read operation) when the pipeline gets triggered is valid.